hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java

```diff
@@ -133,6 +133,15 @@ public CompletableFuture<Message> applyTransaction(
       final TransactionContext trx) {
     final CompletableFuture<Message> applyTransactionFuture =
         new CompletableFuture<>();
+    if (getLifeCycleState().isPausingOrPaused()) {
+      // Statemachine is installation snapshot. Discard any request.
```

Comment:
Is this necessary? Shouldn't Ratis handle making sure we are not applying a transaction while a snapshot is being installed?

Reply:
Ratis will call StateMachine#pause before snapshot install, and then call StateMachine#reinitialize after statemachine reloaded.  In SCM StateMachine#reinitialize, the statemachine will turn itself into RUNNING state.  All this happens concurrently with other Ratis appendEntities request handling.  So we need to check the statemachine state in applyTransaction.  Refer to the crash file attached. 

In OM statemachine implementation, we don't check this statemachine state in applyTransaction. Because all transaction output data go into OMDoubleBuffer, and we stop OMDoubleBuffer before we reload DB, and restart OMDoubleBuffer after that.

Reply:
@ChenSammi , SCM implements notifyInstallSnapshotFromLeader(..).  When the leader does not have the log entries, it sends a install snapshot notification to the follower.  Therefore, it cannot have any appendEntries calls during snapshot installation.  Are we sure the follower is installing a snapshot when it crashes?

Reply:
@szetszwo , there is a misleading message in my previous comment.  Follower is crashed when Ratis call statemachine#applyTransaction through StateMachineUpdater, while SCM statemachine is installing snapshot, not the appendEntities request handling.

Reply:
@ChenSammi , for a follower, applyTransaction is called only if there are appendEntities.  So, we need to understand why this is happening.

BTW, we cannot change applyTransaction to ignore the log.  Otherwise, the states among the servers will diverge silently.

Reply:
Are there other logs indicating that the applyTransaction was running at the same time as, before, or after the snapshot install? Other hypothetical possibilities include:
- Snapshot install finished but the sequence ID generator had stale RocksDB handles when applyTransaction ran.
- RocksDB was closed too early and an applyTransaction operation had not yet finished.

Reply:
With ratis master branch code,  TestStorageContainerManagerHA#testAllSCMAreRunning doesn't crash now.  Will try to add more unit tests to verify whether SCM is safe during install snapshot.

Reply:
@ChenSammi , this is great!  Thanks for checking it.